src/ar_http_iface_cowboy_handler.erl

```diff
@@ -0,0 +1,1142 @@
+									]
+								}
+							end,
+							TX#tx.tags
+						)
+					)};
+				_ ->
+					{ok, JSONBlock} = file:read_file(Filename),
+					{TXJSON} = ar_serialize:dejsonify(JSONBlock),
+					Res = val_for_key(Field, TXJSON),
+					{200, #{}, Res}
+			end
+	end;
+
+%% @doc Share the location of a given service with a peer.
+%% POST request to endpoint /services where the body of the request is a JSON encoded serivce as
+%% specified in ar_serialize.
+do_handle(<<"POST">>, [<<"services">>], Req) ->
+	{BodyBin, _} = read_complete_body(Req),
+	{ServicesJSON} = ar_serialize:jsonify(BodyBin),
+	ar_services:add(
+		whereis(http_services_node),
+		lists:map(
+			fun({Vals}) ->
+				{<<"name">>, Name} = lists:keyfind(<<"name">>, 1, Vals),
+				{<<"host">>, Host} = lists:keyfind(<<"host">>, 1, Vals),
+				{<<"expires">>, Expiry} = lists:keyfind(<<"expires">>, 1, Vals),
+				#service { name = Name, host = Host, expires = Expiry }
+			end,
+			ServicesJSON
+		)
+	),
+	{200, #{}, "OK"};
+
+%% @doc Return the current block hieght, or 500
+do_handle(Method, [<<"height">>], _Req)
+		when (Method == <<"GET">>) or (Method == <<"HEAD">>) ->
+	case ar_node:get_height(whereis(http_entrypoint_node)) of
+		-1 -> {503, #{}, <<"Node has not joined the network yet.">>};
+		H -> {200, #{}, integer_to_binary(H)}
+	end;
+
+%% @doc If we are given a hash with no specifier (block, tx, etc), assume that
+%% the user is requesting the data from the TX associated with that hash.
+%% Optionally allow a file extension.
+do_handle(<<"GET">>, [<< Hash:43/binary, MaybeExt/binary >>], Req) ->
+	case MaybeExt of
+		<< ".", Part/binary >> ->
+			do_handle(<<"GET">>, [<<"tx">>, Hash, <<"data.", Part/binary>>], Req);
+		<<>> ->
+			do_handle(<<"GET">>, [<<"tx">>, Hash, <<"data.html">>], Req);
+		_ ->
+			not_found()
+	end;
+
+%% @doc Catch case for requests made to unknown endpoints.
+%% Returns error code 400 - Request type not found.
+do_handle(_, _, _) ->
+	not_found().
+
+% Cowlib does not yet support status code 208 properly.
+% See https://github.com/ninenines/cowlib/pull/79
+handle208(208) -> <<"208 Already Reported">>;
+handle208(Status) -> Status.
+
+arweave_peer(Req) ->
+	{{IpV4_1, IpV4_2, IpV4_3, IpV4_4}, _TcpPeerPort} = cowboy_req:peer(Req),
+	ArweavePeerPort =
+		case cowboy_req:header(<<"x-p2p-port">>, Req) of
+			undefined -> ?DEFAULT_HTTP_IFACE_PORT;
+			Binary -> binary_to_integer(Binary)
+		end,
+	{IpV4_1, IpV4_2, IpV4_3, IpV4_4, ArweavePeerPort}.
+
+split_path(Path) ->
+	lists:filter(
+		fun(C) -> C /= <<>> end,
+		string:split(Path, <<"/">>, all)
+	).
+
+sendfile(Filename) ->
+	{sendfile, 0, filelib:file_size(Filename), Filename}.
+
+read_complete_body(Req) ->
+	read_complete_body(Req, <<>>).
+
+read_complete_body(UnreadReq, Acc) ->
+	case cowboy_req:read_body(UnreadReq) of
+		{more, Data, ReadReq} ->
+			read_complete_body(ReadReq, <<Acc/binary, Data/binary>>);
+		{ok, Data, ReadReq} ->
+			{<<Acc/binary, Data/binary>>, ReadReq}
```

Comment:
Even though Erlang should be able to handle this efficiently, the standard way of doing this, which is also probably even more efficient, would this:
```
		{more, Data, ReadReq} ->
			read_complete_body(ReadReq, [Data | Acc]);
		{ok, Data, ReadReq} ->
			Body = iolist_to_binary(lists:reverse([Data | Acc]))
			{Body, ReadReq}
```

What do you think about this?

Reply:
I just saw that the binary accumulator approach is in the example code in the Cowboy docs 👍👍